### PR TITLE
LibJS+LibJIT: Move x86_64-specific implementations and prepare for supporting other architectures

### LINKED ISSUE #21669: LibJIT: Support other architectures
I want to preface this by saying that I want to volunteer contributing that feature, both abstracting the current x86 specific implementation and adding a new ARM one. This issue is meant more as a discussion in what direction the project wants to go with that library.

Personally, I see two main approaches:
- LibJIT exposes Assembler, which emits machine code of the host, and the user doesn't have to concern themselves with anything. They use Assembler and it just works™
- LibJIT exposes Assembler, which is an abstract interface and specific Assembler instances per supported architecture. Additionally we could offer a helper which automagically returns the correct Assembler instance based on the host or returns an error about an unsupported architecture. 

The former would be easier in some instances and definitely enough for the current use case (JIT in LibJS), but personally I think the latter would be more flexible (and would make it easier for consumers of the LibJIT library to decide at runtime whether to JIT or just interpret instead of having to dance with preprocessor macros to figure it out).

There could obviously be something I'm not seeing though, so please either correct me or propose different solutions 😄

### PR DESCRIPTION
**Pull request by nrabulinski**
Closes #21669

I renamed the currently x86-specific `Assembler` to `X86_64Assembler` and added a `typedef X86_64Assembler Assembler` assuming the target architecture is x86_64. I also added a new define `JIT_ARCH_SUPPORTED` when LibJIT supports the target architecture.

### CODE REVIEW

File: Userland/Libraries/LibJIT/Assembler.cpp [hunk 1]
```diff
@@ -1,11 +1,23 @@
...
+#include <LibJIT/X86_64/Assembler.h>
 
 namespace JIT {
 
+OwnPtr<Assembler> assembler_for_current_arch([[maybe_unused]] Vector<u8>& output)
+{
+#if ARCH(X86_64)
+    return make<X86_64::X86_64Assembler>(output);
+#else
+    return {};
```
Comment (caoimhebyrne) [inline Userland/Libraries/LibJIT/Assembler.cpp:19]: Could we possibly do a more-descriptive error here at compile time instead, considering we're using the preprocessor here anyways?

Comment (nrabulinski) [reply Userland/Libraries/LibJIT/Assembler.cpp:19]: I could return an error and change the return type to `ErrorOr` but the whole idea was that this function can be called by anyone, regardless if the assembler supports that arch or not, and the caller checks if they got back an Assembler or not

**Comment by awesomekling** [comment]
> On my M1 Macbook test-js runs in 23.5s with those changes applied and in 22.6s without, which makes the virtual dispatch a less than 4% performance decrease. That's emulating x86, obviously.

4% is an unfortunate performance regression!

Perhaps we could make that go away by organizing the code using templates instead of virtuals?

```cpp
template<typename Assembler>
class Compiler<Assembler> {
    ...

	Assembler m_assembler;
};
```

Do you think a design like that would cover everything we want?

**Comment by awesomekling** [comment]
Alternatively, we could use virtuals but still make sure to instantiate the architecture-specific assembler directly in `JS::JIT::Compiler` (via preprocessor macros at the member declaration site)

As long as the leaf assembler classes are marked `final`, the compiler should be able to devirtualize all the calls. And we keep the ability to instantiate any assembler we want.

**Comment by nrabulinski** [comment]
I think both should work fine but I'm wondering if we're giving up any performance by instantiating `Compiler` every time instead of doing it once and keeping the instance around. I'll play with all of those ideas (and finally fix the wasm ci)

File: Userland/Libraries/LibJIT/Assembler.cpp (hunk 1 above)
Comment (ADKaster) [reply Userland/Libraries/LibJIT/Assembler.cpp:19]: If you used `ErrorOr<Assembler>` We could skip the heap indirection. Assuming Assembler remains moveable. Then the compiler could hold an `Optional<Assembler>`

File: Userland/Libraries/LibJS/JIT/Compiler.h [hunk 2]
```diff
@@ -23,104 +21,102 @@ class Compiler {
...
+    Assembler::Reg ARG1 = Assembler::Reg::empty();
+    Assembler::Reg ARG2 = Assembler::Reg::empty();
+    Assembler::Reg ARG3 = Assembler::Reg::empty();
+    Assembler::Reg ARG4 = Assembler::Reg::empty();
+    Assembler::Reg ARG5 = Assembler::Reg::empty();
+    Assembler::Reg RET = Assembler::Reg::empty();
+    Assembler::Reg STACK_POINTER = Assembler::Reg::empty();
+    Assembler::Reg REGISTER_ARRAY_BASE = Assembler::Reg::empty();
+    Assembler::Reg LOCALS_ARRAY_BASE = Assembler::Reg::empty();
+    Assembler::Reg UNWIND_CONTEXT_BASE = Assembler::Reg::empty();
```
Comment (ADKaster) [inline Userland/Libraries/LibJS/JIT/Compiler.h:36]: The move of these from compile-time constants to member variable accesses is very unfortunate. 

Can we do something like:

```
#if ARCH(X86_64)
#include <LibJIT/X86_64/Assembler.h>
using ArchAssembler = X86_64::Assembler;
#elif ARCH ....
#endif
```

And have these generic reg names defined on the arch-specific assembler that we can just delegate to? 
```
struct X86Thingy {
  static constexpr auto GPR0 = 4;
};

using ArchThingy = X86Thingy;


struct GenericThingy {
   static constexpr auto GPR0 = ArchThingy::GPR0; 
};
```

File: Userland/Libraries/LibJS/JIT/Compiler.cpp [hunk 3]
```diff
@@ -20,28 +22,57 @@
...
+Compiler::Compiler(Bytecode::Executable& bytecode_executable)
+    : m_assembler(::JIT::assembler_for_current_arch(m_output))
+    , m_bytecode_executable(bytecode_executable)
+{
+    if (m_assembler.ptr() == nullptr)
+        return;
+
+    // For architectures we support we need to populate all the registers
+    // with their architecture-specific counterparts.
+    if (m_assembler->arch == "x86_64") {
```
Comment (ADKaster) [inline Userland/Libraries/LibJS/JIT/Compiler.cpp:55]: A string? This should be an enum

**Review by ADKaster — COMMENTED** [review]
I'm not sure what the use case is for JITting code for a different architecture than native is other than unit tests? Here's a couple suggestions leaning more towards the compile-time selection of assembler instances.

**Comment by nrabulinski** [comment]
> I'm not sure what the use case is for JITting code for a different architecture than native is other than unit tests?

My thinking was that `LibJIT` may not necessarily be used for JITting, but also as a code emitter for a compiler backend or something, hence my suggestion about making `Assembler` virtual and adding the option to instantiate the correct `Assembler` instance at runtime

File: Userland/Libraries/LibJIT/Assembler.cpp (hunk 1 above)
Comment (nrabulinski) [reply Userland/Libraries/LibJIT/Assembler.cpp:19]: Could we though? We'd need to make Assembler not virtual then, no? Which would then mean changing the whole approach

**Comment by alimpfard** [comment]
Drive-by comment:

> use uintptr_t

Please spell that `FlatPtr`.

**Comment by nrabulinski** [comment]
Completely changed the approach and changed the PR description as per the recent Discord discussion :)

**Pull request merged**